keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -884,19 +891,52 @@ However, as disabling taint-manager is deprecated it is not a concern for this
+unscheduled Pending pods which are Terminating.
+
+#### Marking Scheduled Pending and Terminating pods as Failed by Kubelet
+
+A scheduled pod may also get stuck in a Pending phase (for example due to an
+invalid image reference). In that case, as long as the pod is not Terminating it
+could be a transient issue and Kubelet can may progress after retrying. However,
+once such a Pod (in a Pending phase) is also Terminating, Kubelet is not going
+to attempt starting of the pod. A scheduled Pending pod transitions to the
+Terminating state when the `deletionTimestamp` is set by API server during
+handling of a Delete request, which can be called on the pod by a user or a k8s
+component. The possible scenarios include scheduler preemption or eviction
+triggered by a user via API.
+
+This could be problematic in case of Job pods which have finalizers, because
+such a pod would get stuck in the `Pending` phase. Thus, Job controller currently
+(prior to this KEP) considers such pods as failed, counts it accordingly, and
+removes its finalizer, allowing the API server to complete the pod's deletion.
+
+However, It is important for the pod failure policy to match only against pods
+in terminal phase to avoid race conditions when pod's end state is modified
+(e.g. pod's conditions or exit codes added) while the pod is being matched
+against the policy. In the first iteration of Beta we only require pods to be in
+terminal phase when `podFailurePolicy` is specified (and only in that case).
+This allows to opt-out for the behavior without disabling the feature gate flags
+in case the handling of Pending and Terminating pods is problematic for a Beta user
+(see Issue: [Job controller should wait for Pods to terminate to match the failure policy](https://github.com/kubernetes/kubernetes/issues/113855)).
+
+For the second iteration of Beta (1.27), in order to solve this issue, we plan
+to modify Kubelet to transition such pods (scheduled, Pending and Terminating)
+into the `Failed` phase, allowing the Job controller to count them, match
+against the (optional) pod failure policy and remove their Job finalizers,
+allowing API server to complete deletion. A prototype implementation is prepared
+here: [Mark Pending Terminating pods as Failed by Kubelet](https://github.com/kubernetes/kubernetes/pull/115331).
+
+Note that, the logic could be scoped to only pods with finalizers to save an
+unnecessary API call by Kubelet for the phase transition in case there is no
+finalizer, as in this case the pods are about to be deleted by API server anyway.
+However, we prefer simplicity of the implementation without an additional special
+case if the performance gain is not clear.
+
+One release after the PodDisruptionCondition feature gate graduates to GA
+we plan to simplify Job controller to consider as failed (and count as such)
+pods which are in terminal phase regardless of the fact if the `podFailurePolicy`
+is specified (see [Deprecation](#deprecation)).
```

Comment:
It has to be 2 releases after the kubelet changes. I would just postpone the GA release to 1.29.

Reply:
Updated `kep.yaml` accordingly.